### PR TITLE
New plugin: schneiderWallbox

### PR DESCRIPTION
**Pull request by SebastianRinger**
nymea-plugins-modbus pull request checklist:

- [x] Make sure the pull request's title is of format "Plugin name: Add support for xyz" or "New plugin: Plugin name"

- [x] Did you test the changes on hardware, if not (e.g. absence of required hardware), please mention a person to confirm it has been tested.

- [x] Did you update the plugin's README.md accordingly?

- [ ] Did you update translations (`cd builddir && make lupdate`)?

### CODE REVIEW

File: schneiderWallbox/integrationpluginschneiderwallbox.cpp [hunk 1]
```diff
@@ -0,0 +1,546 @@
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
+*
+* Copyright 2013 - 2020, nymea GmbH
```
Comment (mzanetti) [inline schneiderWallbox/integrationpluginschneiderwallbox.cpp:3]: year out of date (please verify other files too in this regard)

File: schneiderWallbox/integrationpluginschneiderwallbox.cpp [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+    }
+
+    qCDebug(dcSchneiderElectric()) << "Starting network discovery...";
+    NetworkDeviceDiscoveryReply *discoveryReply = hardwareManager()->networkDeviceDiscovery()->discover();
+    connect(discoveryReply, &NetworkDeviceDiscoveryReply::finished, info, [=](){
+        qCDebug(dcSchneiderElectric()) << "Discovery finished. Found" << discoveryReply->networkDeviceInfos().count() << "devices";
+        foreach (const NetworkDeviceInfo &networkDeviceInfo, discoveryReply->networkDeviceInfos()) {
+            qCDebug(dcSchneiderElectric()) << networkDeviceInfo;
+
+            if (!networkDeviceInfo.macAddressManufacturer().contains("TELEMECANIQUE", Qt::CaseSensitivity::CaseInsensitive)) {
```
Comment (mzanetti) [inline schneiderWallbox/integrationpluginschneiderwallbox.cpp:64]: This will cause all devices made by TELEMECANIQUE to appear as such a wallbox. Sounds like it's not enough to filter properly.

Also, are you sure that all models will have mac addresses registered by TELEMATIQUE? Reading up on the history of the TELEMATIQUE - Schneider merger, my personal guess would be that they will use up macs registered by telematique and then continue with macs on their own.

This seems like it needs some more thoughts unless you can confirm that the assumptions made will hold (all macs by TELEMATIQUE will be used for the wallbox and all wallboxes will only use macs by TELEMATIQUE)

File: schneiderWallbox/integrationpluginschneiderwallbox.cpp [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+        info->finish(Thing::ThingErrorNoError);
+    });
+}
+
+
+void IntegrationPluginSchneiderWallbox::setupThing(ThingSetupInfo *info)
+{
+    Thing *thing = info->thing();
+
+    /* Probably not needed.
```
Comment (mzanetti) [inline schneiderWallbox/integrationpluginschneiderwallbox.cpp:94]: Indeed, not needed. if the .json defines only one thing class, you'll never get any other thingclassId here. Anything else would be a bug in nymea core which I though should be caught by the existing auto-tests. You can safely delete this.

File: schneiderWallbox/integrationpluginschneiderwallbox.cpp [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+
+    MacAddress mac = MacAddress(thing->paramValue("mac").toString());
+    if (!mac.isValid()) {
+        info->finish(Thing::ThingErrorInvalidParameter, QT_TR_NOOP("The given MAC address is not valid."));
+        return;
+    }
+    NetworkDeviceMonitor *monitor = hardwareManager()->networkDeviceDiscovery()->registerMonitor(mac);
+
+    uint port = thing->paramValue("port").toUInt();
+    quint16 slaveId = thing->paramValue("slaveId").toUInt();
```
Comment (mzanetti) [inline schneiderWallbox/integrationpluginschneiderwallbox.cpp:117]: Can port and slaveID be changed on the device? Normally they can't be changed, in which case I would just drop the paramTypes and hardcode the values in here... If they can be changed, then indeed, this is required.

File: schneiderWallbox/integrationpluginschneiderwallbox.cpp [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+    SchneiderWallboxModbusTcpConnection *schneiderWallboxTcpConnection = new SchneiderWallboxModbusTcpConnection(monitor->networkDeviceInfo().address(), port, slaveId, this);
+    SchneiderWallbox *schneiderWallbox = new SchneiderWallbox(schneiderWallboxTcpConnection, this);
+    connect(info, &ThingSetupInfo::aborted, schneiderWallbox, &SchneiderWallbox::deleteLater);
+    connect(info, &ThingSetupInfo::aborted, monitor, [monitor, this](){ hardwareManager()->networkDeviceDiscovery()->unregisterMonitor(monitor);});
+
+    // ToDO: Check if this is really a Schneider wallbox.
+
+    quint16 minCurrentLimit = thing->paramValue("minChargeCurrent").toUInt();
+    quint16 maxCurrentLimit = thing->paramValue("maxChargeCurrent").toUInt();
+    thing->setStateMinMaxValues("maxChargingCurrent", minCurrentLimit, maxCurrentLimit);
```
Comment (mzanetti) [inline schneiderWallbox/integrationpluginschneiderwallbox.cpp:127]: Note for myself: Depending on the outcome of the min/max discussion in the .json, this needs to be re-evaluated.

File: schneiderWallbox/integrationpluginschneiderwallbox.cpp [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+        setCpwState(thing, cpwState);
+    });
+
+    connect(schneiderWallboxTcpConnection, &SchneiderWallboxModbusTcpConnection::lastChargeStatusChanged, this, [thing, this](SchneiderWallboxModbusTcpConnection::LastChargeStatus lastChargeStatus){
+        qCDebug(dcSchneiderElectric()) << thing << "Last charge status changed" << lastChargeStatus;
+        setLastChargeStatus(thing, lastChargeStatus);
+    });
+
+    connect(schneiderWallboxTcpConnection, &SchneiderWallboxModbusTcpConnection::stationEnergyChanged, this, [thing](quint32 stationEnergyInWattHours){
+        double stationEnergyInKilowattHours{(static_cast<double>(stationEnergyInWattHours)) / 1000.0};
```
Comment (mzanetti) [inline schneiderWallbox/integrationpluginschneiderwallbox.cpp:180]: I think unless using a *very* old gcc (which wouldn't support the list initializer), the static_cast to double is not needed and the `.0` in the `/1000.0` will do the right thing.

File: debian/nymea-plugin-schneider-wallbox.install.in [hunk 7]
```diff
@@ -0,0 +1,2 @@
+usr/lib/@DEB_HOST_MULTIARCH@/nymea/plugins/libnymea_integrationpluginschneiderWallbox.so
+schneiderWallbox/translations/*qm usr/share/nymea/translations/
```
Comment (mzanetti) [inline debian/nymea-plugin-schneider-wallbox.install.in:2]: This would be the first plugin not following the naming convention of everything being lowercase... I see where it's coming from and it's probably not a big deal... In my experience this stuff will sooner or later break a regexp in some script or something somewhere (thinking docs generators and stuff)...

File: schneiderWallbox/schneiderwallbox.cpp [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+        currentPower += m_stationIntensityPhase2 * 230;
+    }
+    if (m_stationIntensityPhase3 > 0.1) {
+        phaseCount++;
+        currentPower += m_stationIntensityPhase3 * 230;
+    }
+    if (phaseCount != m_phaseCount) {
+        m_phaseCount = phaseCount;
+        emit phaseCountChanged(phaseCount);
+    }
```
Comment (mzanetti) [inline schneiderWallbox/schneiderwallbox.cpp:82]: without having tested it, as I read it, I believe this is slightly problematic:

considering a wallbox and car which both support 3-phase charging:

when it's *not* charging and every phase current is 0, it would result in a phaseCount of 1 and set the state to 1. As soon as it starts charging it would then set it to 3 and set it back to 1 when it's done.

This would cause an optimizer to believe it can only charge on 1 phase while in fact it does 3 phases and make it work with wrong assumptions, forcing it to re-evaluate all of it's output as soon as charging starts.

The way I would write this is:

```
int phaseCount = 0;
if (m_stationIntensityPhase1 > 0.1) phaseCount++;
if (m_stationIntensityPhase2 > 0.1) phaseCount++;
if (m_stationIntensityPhase3 > 0.1) phaseCount++;

// Only update if we actually know the amount of phases
if (phaseCount > 0 && phaseCount != m_phaseCount) {
  m_phaseCount = phaseCount;
  emit ...
}
```

This would cause way less updates to the phasecount state and once it synced up on the wallbox/car combination it would always give the same result until the user really plugs in a different car or uses another cable, which should virtually never happen.

File: schneiderWallbox/integrationpluginschneiderwallbox.cpp [hunk 9]
```diff
@@ -0,0 +1,546 @@
...
+
+    schneiderWallboxTcpConnection->connectDevice();
+
+    connect(schneiderWallbox, &SchneiderWallbox::phaseCountChanged, this, [thing, this](quint16 phaseCount){
+        qCDebug(dcSchneiderElectric()) << thing << "Phase count changed" << phaseCount ;
+        setPhaseCount(thing, phaseCount);
+    });
+
+    /* Probably not needed.
+    connect(schneiderWallbox, &SchneiderWallbox::currentPowerChanged, this, [thing, this](double currentPower){
```
Comment (mzanetti) [inline schneiderWallbox/integrationpluginschneiderwallbox.cpp:205]: if the modbus connection offers a register that gives the corrent currentPower value, indeed I would recommend to delete this code and the custom currentPower calculation in the SchneiderWallbox class to have less confusion what's going on when someone looks at this plugin in 6 months from now.

File: schneiderWallbox/integrationpluginschneiderwallbox.cpp [hunk 10]
```diff
@@ -0,0 +1,546 @@
...
+    });
+    */
+}
+
+
+void IntegrationPluginSchneiderWallbox::postSetupThing(Thing *thing)
+{
+    qCDebug(dcSchneiderElectric()) << "Post setup" << thing->name();
+    if (thing->thingClassId() != schneiderEVlinkThingClassId) {
+        qCWarning(dcSchneiderElectric()) << "Thing class id not supported" << thing->thingClassId();
```
Comment (mzanetti) [inline schneiderWallbox/integrationpluginschneiderwallbox.cpp:219]: same as setupThing(). can safely go away unless you plan to add additional thingClasses to the plugin.

File: schneiderWallbox/integrationpluginschneiderwallbox.cpp [hunk 11]
```diff
@@ -0,0 +1,546 @@
...
+        //m_pluginTimer->start();  apparently not needed anymore.
+    }
+}
+
+void IntegrationPluginSchneiderWallbox::thingRemoved(Thing *thing)
+{
+    qCDebug(dcSchneiderElectric()) << "Removing device" << thing->name();
+    if (m_schneiderDevices.contains(thing)) {
+        disconnect(m_schneiderDevices.value(thing)->modbusTcpConnection(), NULL, thing, NULL);  // This is needed to prevent crash on thing remove.
+        // m_schneiderDevices.value(thing)->modbusTcpConnection()->disconnect(thing);  // This should do exactly the same thing as the line above. Somehow it does not. This line does not prevent the crash.
```
Comment (mzanetti) [inline schneiderWallbox/integrationpluginschneiderwallbox.cpp:247]: yes. `disconnect(QObject*)` is a different thing than `disconnect(QObject *, null, QObject *, null)`. As per Qt documentation, the former will disconnect all signals from the given object (not disconnecting any slots) while the latter will disconnect all signals of the first object from slots of the second object.

In any case. IMO you should be able to drop all of this by changing the line below from

```
m_schneiderDevices.take(thing)->deleteLater()
```
to
```
delete m_schneiderDevice.take(thing):
```
That said, I did not verify if that may change stuff in some other place but it should be ok.

File: schneiderWallbox/integrationpluginschneiderwallbox.cpp [hunk 12]
```diff
@@ -0,0 +1,546 @@
...
+    default:
+        thing->setStateValue("cpwState", "EVSE not available");
+    }
+    thing->setStateValue("pluggedIn", isPluggedIn);
+}
+
+void IntegrationPluginSchneiderWallbox::setLastChargeStatus(Thing *thing, SchneiderWallboxModbusTcpConnection::LastChargeStatus status)
+{
+    switch (status) {
+    case SchneiderWallboxModbusTcpConnection::LastChargeStatusCircuitBreakerEnabled:
```
Comment (mzanetti) [inline schneiderWallbox/integrationpluginschneiderwallbox.cpp:375]: No need to change this, just as a hint for the next time. I normally do something like this instead:
```
QHash<Enum, QString> map = {
  { EnumValue1, "String value for enum value 1" },
  { EnumValue2, "String value for enum value 2" },
};
thing->setStateValue("lastChargeStatus", map.value(status));
```
IMO nicer to read, but again, nothing wrong your code here.

File: schneiderWallbox/integrationpluginschneiderwallbox.cpp [hunk 13]
```diff
@@ -0,0 +1,546 @@
...
+        thing->setStateValue("lastChargeStatus", "Other");
+        break;
+    case SchneiderWallboxModbusTcpConnection::LastChargeStatusUndefined:
+    default:
+        thing->setStateValue("lastChargeStatus", "Undefined");
+    }
+}
+
+void IntegrationPluginSchneiderWallbox::setPhaseCount(Thing *thing, quint16 phaseCount) {
+    thing->setStateValue("phaseCount", phaseCount);
```
Comment (mzanetti) [inline schneiderWallbox/integrationpluginschneiderwallbox.cpp:454]: that looks like boilerplate and can probably just go away :)

File: schneiderWallbox/integrationpluginschneiderwallbox.cpp [hunk 14]
```diff
@@ -0,0 +1,546 @@
...
+        break;
+    case SchneiderWallboxModbusTcpConnection::LastChargeStatusDiDefaultCi:
+        thing->setStateValue("lastChargeStatus", "DI default CI (Conditional Input)");
+        break;
+    case SchneiderWallboxModbusTcpConnection::LastChargeStatusOther:
+        thing->setStateValue("lastChargeStatus", "Other");
+        break;
+    case SchneiderWallboxModbusTcpConnection::LastChargeStatusUndefined:
+    default:
+        thing->setStateValue("lastChargeStatus", "Undefined");
```
Comment (mzanetti) [inline schneiderWallbox/integrationpluginschneiderwallbox.cpp:449]: Oh, another hint. While setting states by name is supported and a perfectly valid thing to do, I would still recommend to prefer using the `setStateValue(const StateTypeId &, ...)` instead if especially when you just have a single thing class.

The reason is that setting states by name is not compile time save. Rename the lastChargeStatus state in the .json and the compiler will *not* complain this way. It would complain if you'd used the stateTypeId instead, thus preventing potential bugs with typos or renames.

File: schneiderWallbox/schneiderwallbox-registers.json [hunk 15]
```diff
@@ -0,0 +1,498 @@
...
+            "registerType": "holdingRegister",
+            "readSchedule": "update",
+            "description": "remoteControllerLifeBit",
+            "defaultValue": "0",
+            "access": "RW"
+        },
+        {
+            "id": "degradedMode",
+            "address": 933,
+            "size": 1,
```
Comment (mzanetti) [inline schneiderWallbox/schneiderwallbox-registers.json:477]: this and the previous one could be put into a block for micro-performance-optimization (reading a block of 2 is faster then reading 2 registers individually). That said, it depends mostly on the device how fast/slow it is to read registers... If the schneider one is very fast, the optimization would be negligible, but we've seen devices where a single read call takes some 300ms, in which case this would a reasonable big optimization.

Again, just a hint, not forcing a change here.

File: schneiderWallbox/schneiderwallbox.cpp [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+        qCDebug(dcSchneiderElectric()) << "Can't set current limit, device is not connected.";
+        return false;
+    }
+
+    qCDebug(dcSchneiderElectric()) << "Setting wallbox max current to" << ampereValue;
+
+    m_chargeCurrentSetpoint = ampereValue;
+    //update();  Don't do update here, because the wallbox does not react very fast to modbus writes.
+    if (m_errorOccured) {
+        return false;
```
Comment (mzanetti) [inline schneiderWallbox/schneiderwallbox.cpp:242]: Hmm.. I can see how you got there. The first intention was probably to directly write the register but then it turned out the whole update() is too slow and blabla... Can't tell you what the way to go is (perhaps what you do now is the best thing to do) but what strikes me here is that you seem to evaluate an outdated m_errorOccurred... Perhaps even that is ok (I can't test it without having such a wallbox) but please take a minute to think about what exactly happens and the corner cases it could run into the effects of return false here are (afaict it will at least cause the ThingActionInfo to finish() with an error, even though it might actually succeed...

**Review by mzanetti — CHANGES_REQUESTED** [review]
Nice pull request. Thank you!

I've some suggestions for improvements but overall it looks pretty good already.

File: schneiderWallbox/integrationpluginschneiderWallbox.json [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+                        {
+                            "id": "e7061cb3-27f3-4981-8672-2e4f16e87a43",
+                            "name": "totalEnergyConsumed",
+                            "displayName": "Total energy consumed",
+                            "displayNameEvent": "Total energy consumption changed",
+                            "type": "double",
+                            "unit": "KiloWattHour",
+                            "defaultValue": 0,
+                            "suggestLogging": true,
+                            "cached": false
```
[the remark this answers is not in the document]
Comment (mzanetti) [reply schneiderWallbox/integrationpluginschneiderWallbox.json:158]: Oh, instead, you probably want to disable logging on it (note that the suggestLogging) will act on the generic log DB. It will still be logged by nymea-experience-plugin-energy in a nice and sampled way, regardless of what you set here.

**Comment by mzanetti** [comment]
Hi @SebastianRinger ,

We would really love to include your contributions into nymea, however, the issue is that the way you're filing the pull requests won't work out. If you're throwing half-baked pull requests at us which aren't ready for actual use but mostly just proof of concepts, this ends up being more work on our side that we'd have to just do it ourselves. Please note that the biggest amount of work in a software project is not the initial bit of typework (that's really the easy and quick part of it), but it's the last 20% required to polish all the details, test things in detail, and then care for the maintenance.

Given we don't have access to the hardware, it's quite impossible for us to complete those last 20%, especially without you responding to comments or questions at all. Especially it feels like you're just trying to offload some work to us to complete the integration so you can use it again when it's done.

So, please complete the pull requests and respond to reviews, bringing the plugin to a production-ready state or we'll sadly have to discard them.

**Comment by mzanetti** [comment]
Closing due to lack of activity
